rfc/rfc-89/rfc-89.md

```diff
@@ -0,0 +1,379 @@
+
+Apache Hudi natively supports multi-versioning and version rollback capabilities. For Bucket Rescale operation, 
+we must also support Bucket Rescale rollbacks, even rolling back multiple completed Bucket Rescale operations at once. 
+This requires tracking the instant and corresponding config version for each Bucket Rescale action. 
+During rollback, both the data and config must be reverted to ensure consistency between the configuration and the data state.
+
+So that we need to maintain config versions alongside commit instants which is multiple versions of hashing_config (multi-versioned configs). 
+This necessitates implementing mechanisms for config creation, update, rollback, cleanup and loading
+
+About Hashing_Config Creation
+
+The initial version *must and can only* be created via DDL. During the DDL phase, hoodie will create a 
+`00000000000000000.hashing_config` for the first time.
+
+About Hashing_Config Update
+
+A new version can only be updated via `call partitionBucketIndexManager(overwrite => 'new-expressions', dry-run => 'false')`, which is 
+a replace-commit operation. 
+
+The transactional relationship between the replace-commit operation and hashing_config updates will be explained in the 
+PartitionBucketIndexManager section later.
+
+*Note:* Before updating the hashing_config, users must manually stop all ingestion jobs.
+
+```text
+Firstly DDL phase creates initial version 
+ls .hoodie/.hashing_meta/
+.hoodie/.hashing_meta/00000000000000000.hashing_config
+{
+   "rule": "rule-engine",
+   "expressions": "expression1,bucket-number1;expression2,bucket-number2",
+   "default_bucket_number": "default-bucket-number"
+}
+
+Secondly call PartitionBucketIndexManager(overwrite => 'expression3,bucket-number3;expression1,bucket-number1;expression2,bucket-number2', 'dry-run' = 'false')
+ls .hoodie/.hashing_meta/
+.hoodie/.hashing_meta/00000000000000000.hashing_config (deprecated)
+.hoodie/.hashing_meta/20250303095546020.hashing_config (used)
+{
+   "rule": "rule-engine",
+   "expressions": "expression3,bucket-number3;expression1,bucket-number1;expression2,bucket-number2",
+   "default_bucket_number": "default-bucket-number"
+}
+```
+
+The expression written at the front has a higher priority, and the expression has a higher priority than the default_bucket_number.
+
+About Hashing_Config Rollback
+
+The Bucket Rescale operation is essentially an insert overwrite action and can seamlessly integrate with Hudi's rollback mechanism. 
+One simplification here is that when rolling back a Bucket Rescale, the corresponding replace-commit.hashing_config is not immediately deleted. 
+Instead, its cleanup is deferred to the version cleanup process (discussed later).
+
+This approach achieves two goals:
+1. Minimizes modifications to the rollback service logic.
+2. Lazy cleanup does not impact config loading, as Hudi always loads the latest hashing_config associated with committed replace commit.
```

Comment:
Consider the following scenario(lazy clean)

001.repalce.inflight        generate  001.hashing_config
002.commit
003.commit
004.commi

hudi will not load 001.hashing_config as 001.repalce is not committed

then we do clean
clean Action:   rollback and delete 001.repalce.inflight 

the commitTime of Hasing_config is 001 which samller than the earliest finished Instant 002.commit.
what’s the state of 001.hashing_config now.

Reply:
The `001.hashing_config` should be removed in the clean process.

Reply:
Updated. We will rollback hashing config during rollback insert overwrite(Bucket rescale action)